### ISSUE TITLE
error: expected one of `)` or `,`, found `[` using empty slice &[][..] in if let expression

### ISSUE THREAD

**Issue by ncalexan** [rust-lang/rust#38892]
With the following input, I get an unexpected error.  (I'm testing locally and also using https://play.rust-lang.org; links below.)
```rust
fn main() {
    let x = Ok((1, &[][..]));
    if let Ok((1, &[][..])) = x {
        true
    } else {
        false
    }
}
```
results in an error parsing (lexing?) `&[][..]` in the `if let` expression:
```
error: expected one of `)` or `,`, found `[`
...
```
(See https://play.rust-lang.org/?gist=95b4c8fbaba1233075155fd33641d13a&version=nightly&backtrace=0.)

If I extract a temporary variable, like the following:
```rust
fn main() {
    let t: &[i32] = &[][..];
    let x: Result<(i32, &[i32]), &str> = Ok((1 as i32, t));
    if let Ok((1, t)) = x {
        true
    } else {
        false
    };
}
```
then everything compiles, but I get a spurious unused variable warning (again in the `if let` expression):
```
error: expected one of `)` or `,`, found `[`
...
```
(See https://play.rust-lang.org/?gist=90653e50596f3f258f29b27a5cf44d69&version=nightly&backtrace=0.)

I see no differences running stable or nightly builds.

**Comment by arielb1** [comment]
`if let` expects a [pattern](https://doc.rust-lang.org/book/patterns.html), not an expression.

**Comment by ncalexan** [comment]
> if let expects a pattern, not an expression.

I understand this (I think) so let me change my question: why is the literal `&[][..]` not accepted in my pattern `Ok((1, &[][..]))`?

**Comment by arielb1** [comment]
Because the pattern syntax is fairly restricted - you can't specify arbitrary expressions in it.

The empty slice pattern is specified as just `[]` (for both `[T; 4]` arrays and `[T]` slices), so you could also have used `Ok((1, &[]))`, except that due to some bugs these kinds of patterns are unstable.

**Comment by ncalexan** [comment]
> Because the pattern syntax is fairly restricted - you can't specify arbitrary expressions in it.

OK, I was not aware of this.

@arielb1 can you also comment on the spurious unused variable warning I see when I hoist the arbitrary expression out of the pattern expression?  Is that expected?  It certainly seems wrong to me, but perhaps used is something equivalent to lvalue usage, where I'm doing something equivalent to rvalue usage?  That is, I should declare my value as const or similar to avoid the unused variable warning?

**Comment by arielb1** [comment]
The pattern `Ok((1, t))` binds the variable `t`. You can then access `t` within the match, as in:
```Rust
    if let Ok((1, t)) = x {
        println!("{}", t);
        true
    } else {
        false
    };
```

Therefore, the unused variable warning is very much correct - if you are not interested in `t`, you should not bind it.

**Comment by ncalexan** [comment]
> The pattern Ok((1, t)) binds the variable t.

Ah, I see.  Thanks for all your help, @arielb1!
